cpp/src/arrow/util/hashing.h

```diff
@@ -300,36 +304,57 @@ class HashTable {
 
   bool NeedUpsizing() const {
     // Keep the load factor <= 1/2
-    return n_filled_ * 2U >= size_;
+    return size_ * kLoadFactor >= capacity_;
   }
 
-  void Upsize(uint64_t new_size) {
-    assert(new_size > size_);
-    uint64_t new_mask = new_size - 1;
-    assert((new_size & new_mask) == 0);  // it's a power of two
+  Status UpsizeBuffer(uint64_t capacity) {
+    RETURN_NOT_OK(entries_builder_.Resize(capacity));
+    entries_ = entries_builder_.mutable_data();
+    memset(entries_, 0, capacity * sizeof(Entry));
```

Comment:
This caused the following warning (this is treated as error in debug build) with g++ 8.3.0.

```text
../src/arrow/util/hashing.h:313:11: error: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct arrow::internal::HashTable<arrow::internal::ScalarMemoTable<nonstd::sv_lite::basic_string_view<char>, arrow::internal::HashTable>::Payload>::Entry'; use assignment or value-initialization instead [-Werror=class-memaccess]
     memset(entries_, 0, capacity * sizeof(Entry));
     ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/arrow/util/hashing.h:197:10: note: 'struct arrow::internal::HashTable<arrow::internal::ScalarMemoTable<nonstd::sv_lite::basic_string_view<char>, arrow::internal::HashTable>::Payload>::Entry' declared here
   struct Entry {
          ^~~~~
cc1plus: all warnings being treated as errors
```

Reply:
May have to suppress this warning here

Reply:
I've fixed this by #5001.